src/java.base/share/classes/java/util/concurrent/locks/AbstractQueuedLongSynchronizer.java

```diff
@@ -277,6 +277,26 @@ private static void signalNextIfShared(Node h) {
         }
     }
 
+    /**
+     * Invokes acquire for as long as it throws exceptions
+     * @param node which to reacquire
+     * @param arg the acquire argument
+     */
+    private final void reacquire(Node node, long arg) {
+        for(boolean first = true;;) {
+            try {
+                acquire(node, arg, false, false, false, 0L);
+                break;
+            } catch (Error | RuntimeException ex) {
```

Comment:
@DougLea @AlanBateman The first question is if this set of exception are the way to go, or we should go with Throwable

Reply:
One option is to record the first encountered exception instance, which can then be rethrown once acquire succeeds?

Reply:
Object.wait and Condition.await are unusual in that they must reacquire they continue, even if they throw. So might not be crazy for the cases where it does recover (I assume some cases will never recover and it will retry indefinitely).

Reply:
@DougLea What do you think about storing the initial exception and rethrowing it on an eventual retry success?